src/charm.py

```diff
@@ -214,16 +217,44 @@ def _on_start(self, event) -> None:
 
             self.app_peer_data["db_initialised"] = "True"
 
+    def is_unit_in_replica_set(self) -> bool:
+        """Check if the unit is in the replica set."""
+        with MongoDBConnection(self.mongodb_config) as mongo:
+            try:
+                replset_members = mongo.get_replset_members()
+                return self.get_hostname_by_unit(self.unit.name) in replset_members
+            except NotReadyError as e:
+                logger.error(f"{self.unit.name}.is_unit_in_replica_set NotReadyError={e}")
+            except PyMongoError as e:
+                logger.error(f"{self.unit.name}.is_unit_in_replica_set PyMongoError={e}")
+
+    def _on_stop(self, event) -> None:
```

Comment:
I believe you could move this logic to the `storage_detaching` hook which is the first hook to be called on scaledown. And you already have some logic related to the scaledown there.

And do not use the `stop` hook (following conversation with @marcoppenheimer about upgrades, this hook gets called on a juju upgrade action - so better leave it for now)

Reply:
I am not sure about using `storage_detaching`, the usage of `stop` hook was suggested by juju devs. The idea is to block deletion of unit and the `stop` is the correct place to do it

Reply:
`stop` comes later in the teardown process, `storage_detaching` is the entry point of teardown (you can confirm this by using `jhack utils tail -l 30 -d` then deleting the unit and watch the lifecycle) so it doesn't hurt to move the logic there imo. For reference: https://github.com/canonical/mongodb-operator/blob/main/src/charm.py#L155

Reply:
we still need to be sure what by the time app leader receives `RelationDepartedEvent` for a unit, the unit  and its pod is not deleted.